src/openms/source/CHEMISTRY/TheoreticalSpectrumGeneratorXLinks.cpp

```diff
@@ -0,0 +1,1271 @@
+      }
+
+      if (add_losses_) // add loss peaks (slow)
+      {
+        Size i = xlink_pos_A+1;
+        for (; i < peptideA.size(); ++i)
+        {
+          const AASequence ion = peptideA.getPrefix(i);
+          addXLinkLosses_(spec_alpha, ion, AASequence::fromString(""), cross_link.cross_linker_mass, res_type, charge, intensity, ion_type);
+        }
+      }
+    }
+    else // if (res_type == Residue::XIon || res_type == Residue::YIon || res_type == Residue::ZIon)
+    {
+      if ((!add_isotopes_) || max_isotope_ < 3) // add single peaks (and maybe a second isotopic peak)
+      {
+        double mono_weight(Constants::PROTON_MASS_U * charge + cross_link.cross_linker_mass);
+        if (xlink_pos_B == peptideA.size()+1 && peptideA.hasCTerminalModification())
+        {
+          mono_weight += peptideA.getCTerminalModification()->getDiffMonoMass();
+        }
+
+        Size i = peptideA.size() - xlink_pos_B - 1;
+        if (i < peptideA.size())
+        {
+          mono_weight += peptideA.getSuffix(i).getMonoWeight(Residue::Internal);
+        }
+        for (Size k = peptideA.size() - i - 1; k > 0; --k)
+        {
+          i++;
+          mono_weight += peptideA[k].getMonoWeight(Residue::Internal);
+          double pos(mono_weight);
+          switch (res_type)
+          {
+            case Residue::XIon: pos = (pos + Residue::getInternalToXIon().getMonoWeight()) / charge; break;
+            case Residue::YIon: pos = (pos + Residue::getInternalToYIon().getMonoWeight()) / charge; break;
+            case Residue::ZIon: pos = (pos + Residue::getInternalToZIon().getMonoWeight()) / charge; break;
+            default: break;
+          }
+          addPeak_(spec_alpha, pos, intensity, res_type, i-1, charge, ion_type);
+          if (add_isotopes_ && max_isotope_ == 2) // add second isotopic peak with fast method, of only two peaks are asked for
+          {
+            pos += Constants::C13C12_MASSDIFF_U / charge;
+            addPeak_(spec_alpha, pos, intensity, res_type, i-1, charge, ion_type);
+          }
+        }
+      }
+      else // add isotope clusters (slow)
+      {
+        Size i = peptideA.size() - xlink_pos_B;
+        for (; i < peptideA.size(); ++i)
+        {
+          const AASequence ion = peptideA.getSuffix(i);
+          addIsotopeCluster_(spec_alpha, ion, AASequence::fromString(""), cross_link.cross_linker_mass, res_type, charge, intensity, ion_type);
+        }
+      }
+
+      if (add_losses_) // add loss peaks (slow)
+      {
+        Size i = peptideA.size() - xlink_pos_B;
+        for (; i < peptideA.size(); ++i)
+        {
+          const AASequence ion = peptideA.getSuffix(i);
+          addXLinkLosses_(spec_alpha, ion, AASequence::fromString(""), cross_link.cross_linker_mass, res_type, charge, intensity, ion_type);
+        }
+      }
+    }
+    return;
+  }
+
+  void TheoreticalSpectrumGeneratorXLinks::addAbundantImmoniumIons(RichPeakSpectrum & spec, const AASequence& peptide) const
```

Comment:
this just looks like duplicate code from `src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp` - can you please check whether that code could be merged?